### PR TITLE
[api][web][SIMS-#233] Sync BCeID Account Info

### PR DESCRIPTION
**Pull request by fwpushan**
-------
1. Create new method to fetch institution using userName
2. Create sync flow
---

### CODE REVIEW

File: sources/packages/api/src/services/institution/institution.service.ts [hunk 1]
```diff
@@ -94,24 +94,18 @@ export class InstitutionService extends RecordDataModelService<Institution> {
...
-      );
-      throw new InternalServerErrorException(
-        `Unable to fetch account information from BCeID`,
-      );
-    }
+  async getInstitute(userInfo: { userName: string }): Promise<Institution> {
+    const query = this.repo
+      .createQueryBuilder("institution")
+      .leftJoinAndSelect("institution.users", "users");
+    const institution = await query
```
Comment (andrewsignori-aot) [inline sources/packages/api/src/services/institution/institution.service.ts:101]: Can you please elaborate on why we need to break the query into two objects here?

File: sources/packages/api/src/services/institution/institution.service.ts [hunk 2]
```diff
@@ -94,24 +94,18 @@ export class InstitutionService extends RecordDataModelService<Institution> {
...
-    if (account == null) {
-      // Due to any connection or any other integration error retrieval of account info failed.
-      this.logger.error(
-        "Account information could not be retrieved from BCeID",
-      );
-      throw new InternalServerErrorException(
-        `Unable to fetch account information from BCeID`,
-      );
-    }
+  async getInstitute(userInfo: { userName: string }): Promise<Institution> {
```
Comment (andrewsignori-aot) [inline sources/packages/api/src/services/institution/institution.service.ts:97]: Can this method be named as something like getInstitutionByUserName and have the parameter as userName only?
`async getInstitutionByUserName(userName: string): Promise<Institution>`

File: sources/packages/api/src/services/institution/institution.service.ts [hunk 3]
```diff
@@ -157,13 +151,25 @@ export class InstitutionService extends RecordDataModelService<Institution> {
     return await this.save(institution);
   }
 
+  async syncInstitution(userInfo: UserInfo): Promise<void> {
+    const account = await this.bceidService.getAccountDetails(
+      userInfo.idp_user_name,
+    );
+    const institutionEntity = await this.getInstitute(userInfo);
+    institutionEntity.guid = account.institution.guid;
```
Comment (andrewsignori-aot) [inline sources/packages/api/src/services/institution/institution.service.ts:159]: Please consider not updating the institution guid and, please add validation to check if the institution guid on account object is the same institution guid retrieved from DB to be updated.

File: sources/packages/api/src/services/institution/institution.service.ts [hunk 4]
```diff
@@ -157,13 +151,25 @@ export class InstitutionService extends RecordDataModelService<Institution> {
...
+    institutionEntity.legalOperatingName = account.institution.legalName;
+    await this.save(institutionEntity);
+    const user = await this.userService.getUser(userInfo.userName);
```
Comment (andrewsignori-aot) [inline sources/packages/api/src/services/institution/institution.service.ts:162]: If the user is already loaded in the institutionEntity, please consider changing the user object inside institutionEntity and executing only one save operation.

File: sources/packages/api/src/services/institution/institution.service.ts [hunk 5]
```diff
@@ -157,13 +151,25 @@ export class InstitutionService extends RecordDataModelService<Institution> {
...
   async institutionDetail(userInfo: UserInfo): Promise<InstitutionDetailDto> {
     const account = await this.bceidService.getAccountDetails(
       userInfo.idp_user_name,
     );
-    const institutionEntity = await this.repo.findOne({
-      guid: account.institution.guid,
-    });
+    const institutionEntity = await this.getInstitute(userInfo);
 
     const user = await this.userService.getUser(userInfo.userName);
```
Comment (andrewsignori-aot) [inline sources/packages/api/src/services/institution/institution.service.ts:175]: I think that I missed it in the previous PR, but I believe that the conversion to DTO should happen in the controller layer or anywhere outside this service class.

File: sources/packages/api/src/services/institution/institution.service.spec.ts [hunk 6]
```diff
@@ -1,13 +1,62 @@
...
+    legalAuthorityPhone: faker.phone.phoneNumber(),
+  };
+  institution.regulatingBody = "ICBC";
+  institution.users = [user];
+  await service.save(institution);
+
+  return [institution, user];
+};
 
 describe("InstitutionService", () => {
```
Comment (andrewsignori-aot) [inline sources/packages/api/src/services/institution/institution.service.spec.ts:56]: This service is now interacting with the database, are we going to keep those "Unit Tests" executing always to it should be good to skip them?

**Review by andrewsignori-aot — COMMENTED** [review]
Please take a look in the comments.

File: sources/packages/api/src/services/institution/institution.service.ts (hunk 3 above)
Comment (fwpushan) [reply sources/packages/api/src/services/institution/institution.service.ts]: DONE

File: sources/packages/api/src/services/institution/institution.service.ts (hunk 4 above)
Comment (fwpushan) [reply sources/packages/api/src/services/institution/institution.service.ts]: The egar loading working, so FIXED

File: sources/packages/api/src/services/institution/institution.service.ts (hunk 5 above)
Comment (fwpushan) [reply sources/packages/api/src/services/institution/institution.service.ts:193]: I like mapper concept, but that require additional learning, so skip this for current works. Mapper could be part of entity.

Comment (fwpushan) [reply sources/packages/api/src/services/institution/institution.service.ts:191]: This code is outdated, I don't know why git is not reflecting that

Comment (fwpushan) [reply sources/packages/api/src/services/institution/institution.service.ts:191]: I don't agree for controller, but I agree there should be a mapping layer.

Comment (fwpushan) [reply sources/packages/api/src/services/institution/institution.service.ts:191]: But mapping logic is there from the beginning, in this PR there is little change on that portion. Refactoring them will be totally change the task estimation.

File: sources/packages/api/src/services/institution/institution.service.spec.ts (hunk 6 above)
Comment (fwpushan) [reply sources/packages/api/src/services/institution/institution.service.spec.ts:56]: I will include this as integration test later.

Comment (fwpushan) [reply sources/packages/api/src/services/institution/institution.service.spec.ts:56]: Test skipped

File: sources/packages/api/src/services/institution/institution.service.ts (hunk 2 above)
Comment (fwpushan) [reply sources/packages/api/src/services/institution/institution.service.ts]: DONE

File: sources/packages/api/src/services/institution/institution.service.ts (hunk 1 above)
Comment (fwpushan) [reply sources/packages/api/src/services/institution/institution.service.ts]: DONE

**Pull request merged**